### DISCUSSION TITLE
How to fix the end point of the rod in a determined position?

### DISCUSSION THREAD

**Discussion by Edmund-Wang0546** [GazzolaLab/PyElastica#197] [Q&A]
I hope to get the curve of a Flexible rod with 2 fixed ends, like:
![1](https://user-images.githubusercontent.com/113122422/192674097-b010e40a-10dd-4188-910b-db3b59c807dc.png)
 The start point could be fixed by setting 'start' in straight_rod class and adding 'OneEndFixedBC' with 'constrained_position_idx=(0,)'. **But how to fix the end point of the rod in a determined position?**

**Comment by armantekinalp** [comment]
You can use `GeneralConstrain` and pass the index you want to constrain. https://github.com/GazzolaLab/PyElastica/blob/master/elastica/boundary_conditions.py#L262

**Reply by Edmund-Wang0546** [reply]
Thanks for your reply. Yes, as the example shows, we can constrain the element by 'GeneralConstrain' and index, like constrain the end of rod using 'GeneralConstraint' and 'constrained_position_idx=(-1,)'. But how to set end element's **position**? Like constrain the end (index=-1) at position (1,0,0) in coordinate system.

**Reply by armantekinalp** [reply]
What is your initial condition? Is the rod tip at position (1,0,0) ?

**Reply by Edmund-Wang0546** [reply]
Thank you. This is a very simple example. We just hope to verify the deformation of a flexible rod. For example, the rod's length is 1.2, start element is fixed at (0,0,0) and end element is fixed at (1,0,0). These would never change during the simulation. Thus, no motion is generated during the simulation. I just hope to see the curve of the rod.

**Reply by armantekinalp** [reply]
Do you know the rod node positions? or do you expect PyElastica to compute rod node positions based on start, end position and total length? If it is the second case this is not possible because there can be many different deformations can happen. I would suggest to start with a straight rod and apply a force in `-x` direction at the tip to move rod end from `(1.2, 0, 0)` to `(1,0,0)`.

**Reply by cuicw** [reply]
I also encountered a similar problem. I input the curve configuration of the flexible rod as a position into CosseratRod.straight _ rod, and the specific input is as follows : position = positions1.T,, but the following error pops up : 
AssertionError : 
Not equal to tolerance rtol = 1e-07, atol = 2.22045e-12 
Rod normal and tangent are not perpendicular to each other ! 
Mismatched elements : 2 / 3 ( 66.7 % ) 
Max absolute difference : 0.57071637 
Max relative difference : inf 
x : array ( [ -4.471247e-001, -5.707164e-001, 1.750243e-306 ] ) 
y : array ( 0 )

**Reply by cuicw** [reply]
I already know the curve configuration of the flexible rod ( initial point, end point and rod length ), and the position information of each node is also calculated and discretized by the theoretical formula

**Reply by armantekinalp** [reply]
This is a different issue. Please open a new issue. 

Short answer: Provided rod normal is not perpendicular to the rod tangents. Make sure you provide correct normal vector. If you rod does not lie on a plane such that there is not a single normal vector, then you should pass directors to the rod initializer. See the [documentation](https://docs.cosseratrods.org/en/latest/api/rods.html) .
